### PR TITLE
fix #37: pass previous exception code instead of null to the second argument of the exception's constructor

### LINKED ISSUE #37: PHP 8.1 compat: type error in exception constructor
### Bug Report

|    Q        |   A
|------------ | ------
| Version(s)  | 2.11.0-2.12.1

#### Summary
On PHP 8.1 throwing exception for iterator that  couldn't be created ends up with type error.

PHP 8.1 does not accept null as default parameter for exception `$code` parameter when previous exception is passed to the new exception constructor.

This usage appears multiple times in the codebase.

#### Current behavior

```
TypeError: Exception::__construct(): Argument #2 ($code) must be of type int, null given
    in /app/vendor/laminas/laminas-paginator/src/Paginator.php:664
```

https://github.com/laminas/laminas-paginator/blob/15922abdf361f15ca35e585214c3f7091b88f5ee/src/Paginator.php#L663-L665

#### How to reproduce

Dummy adapter that would throw on attempt to get items would trigger the problem code path in Paginator

```
use Laminas\Paginator\Adapter;

class TestDummy implements AdapterInterface
{
    public function getItems($offset, $itemCountPerPage)
    {
        throw new ExpectedException();
    }
}
```

#### Expected behavior

Paginator runtime exception is thrown with previous exception correctly set

### PR DESCRIPTION
**Pull request by franzose**
|    Q          |   A
|-------------- | ------
| Documentation | yes/no
| Bugfix        | **yes**
| BC Break      | yes/no
| New Feature   | yes/no
| RFC           | yes/no
| QA            | yes/no

### Description

This PR aims to fix #37. A test is included just in case.

### CODE REVIEW

**Comment by Ocramius** [comment]
I adjusted the target branch to `2.12.x` - can we rebase on top of it? Alternatively, I can do it for you before merging.

File: src/Paginator.php [hunk 1]
```diff
@@ -661,7 +661,7 @@ public function getIterator()
         try {
             return $this->getCurrentItems();
         } catch (Throwable $e) {
-            throw new Exception\RuntimeException('Error producing an iterator', null, $e);
+            throw new Exception\RuntimeException('Error producing an iterator', 0, $e);
```
Comment (Xerkus) [inline src/Paginator.php]: Note that it is not the only occurrence of null as second parameter. Rest of them would need to be handled as well.

**Review by Xerkus — CHANGES_REQUESTED** [review]

File: src/Paginator.php (hunk 1 above)
Comment (franzose) [reply src/Paginator.php]: Also, it seems to me that unit tests are useless in this case. What do you think?

Comment (Ocramius) [reply src/Paginator.php]: Yeah, I think reducing the psalm baseline suffices: `vendor/bin/psalm --update-baseline` should do that

Comment (franzose) [reply src/Paginator.php]: @Xerkus I've looked through classes, this is the only place.

Comment (franzose) [reply src/Paginator.php]: @Ocramius the branch has been rebased on top of `2.12.x`.

Comment (Xerkus) [reply src/Paginator.php]: > @Xerkus I've looked through classes, this is the only place.

Indeed. My bad. Used wrong grep regex.

File: src/SerializableLimitIterator.php [hunk 2]
```diff
@@ -18,17 +18,13 @@ class SerializableLimitIterator extends LimitIterator implements Serializable, A
 {
     /**
      * Offset to first element
-     *
-     * @var int
      */
-    private $offset;
+    private int $offset;
```
Comment (Xerkus) [inline src/SerializableLimitIterator.php]: This is unrelated change.

Comment (franzose) [reply src/SerializableLimitIterator.php]: Sorry, it might have come after rebasing :/

File: src/Paginator.php [hunk 3]
```diff
@@ -661,7 +661,7 @@ public function getIterator()
         try {
             return $this->getCurrentItems();
         } catch (Throwable $e) {
-            throw new Exception\RuntimeException('Error producing an iterator', null, $e);
+            throw new Exception\RuntimeException('Error producing an iterator', (int) $e->getCode(), $e);
```
Comment (Xerkus) [inline src/Paginator.php:664]: Type cast is not needed since `getCode()` return type is `int`

File: src/SerializableLimitIterator.php (hunk 2 above)
Comment (Xerkus) [reply src/SerializableLimitIterator.php]: Now you removed the typehint.

Sorry for so much hassle for such a simple change

**Review by Xerkus — CHANGES_REQUESTED** [review]
One small change please.

**Comment by Xerkus** [comment]
Also, If you want to squash commits into one, here is a suggestion:

Assuming origin remote is your fork and upstream points to laminas:
```
git checkout fix/37/type-error-in-get-iterator
git reset --soft upstream/2.12.x
git commit -s -m ""
git push --force origin fix/37/type-error-in-get-iterator
```

File: src/Paginator.php (hunk 3 above)
Comment (franzose) [reply src/Paginator.php:664]: Psalm complains if there is no type cast.

**Review by Xerkus — APPROVED** [review]

**Comment by Ocramius** [comment]
Thanks @franzose!

**Comment by franzose** [comment]
Thanks for applying the fix! And sorry for this back-and-forth stuff. Next time should be better prepared.

**Comment by Ocramius** [comment]
@franzose it's like this every time, no worries 👍

**Pull request merged**